Python 3.14: PEP-784 compression.zstd

Rogdham opened the pull request:
Second PR about PEP-784 in stdlib ([first PR](#13992)):
 - Add type hints for `_zstd`
 - Add type hints for `compression.zstd`

Rogdham commented:
@emmatyping: if you want to give try this out, or simply comment on the use of `Mapping[CompressionParameter, int]` and `Mapping[DecompressionParameter, int]` as type hints for the `options` parameter. Any feedback would be appreciated!

emmatyping commented:
> Note that I typed the options as Mapping[CompressionParameter, int] / Mapping[DecompressionParameter, int], while a Mapping[int, int] is technically accepted.

I think we should type that it accepts `Mapping[int, int]` as well. I consider that a supported usage.

emmatyping commented on stdlib/_zstd.pyi:
```diff
@@ -0,0 +1,89 @@
+    def __init__(
+        self, level: int | None = ..., options: _OptionsCompress | None = ..., zstd_dict: ZstdDict | None = ...
+    ) -> None: ...
+    def compress(self, /, data: ReadableBuffer, mode: Literal[0, 1, 2] = ...) -> bytes: ...
```
I think it might be useful to express these in terms of the `ZstdCompressor` attributes. It will make it easier to understand I think? If you think it makes the signature too busy then I suppose it is fine as is, but I want discourage passing bare integers.

Rogdham replied:
I believe the below is [invalid](https://typing.python.org/en/latest/spec/literal.html#legal-and-illegal-parameterizations):

```py
Literal[CONTINUE, FLUSH_BLOCK, FLUSH_FRAME]
```

But maybe I can try to do something like the following:

```py
_ZstdCompressorContinueType = Literal[0]
_ZstdCompressorFlushBlockType = Literal[1]
_ZstdCompressorFlushFrameType = Literal[2]

@final
class ZstdCompressor:
    CONTINUE: Final[_ZstdCompressorContinueType] = 0
    FLUSH_BLOCK: Final[_ZstdCompressorFlushBlockType] = 1
    FLUSH_FRAME: Final[_ZstdCompressorFlushFrameType] = 2
    def compress(self, /, data: ReadableBuffer, mode: _ZstdCompressorContinueType | _ZstdCompressorFlushBlockType | _ZstdCompressorFlushFrameType = ...) -> bytes: ...
```

Is it what you meant?

emmatyping replied:
Okay, I think that is fine then.

emmatyping commented on stdlib/_zstd.pyi:
```diff
@@ -0,0 +1,95 @@
+def get_frame_info(frame_buffer: ReadableBuffer) -> tuple[int, int]: ...
+def get_frame_size(frame_buffer: ReadableBuffer) -> int: ...
+def get_param_bounds(parameter: int, is_compress: bool) -> tuple[int, int]: ...
+def set_parameter_types(c_parameter_type: type[Any], d_parameter_type: type[Any]) -> None: ...
```
Yeah these should be specifically `type[CompressionParameter]` and `type[DecompressionParameter]`.
```suggestion
def set_parameter_types(c_parameter_type: type[CompressionParameter], d_parameter_type: type[DecompressionParameter]) -> None: ...
```

Rogdham replied:
If my understanding is correct, this functions take any types, and will set valid types to be used as keys (on top of `int`) for the `options` parameters.

In the implementation of `zstd` a single call is done explicitly with `CompressionParameter` and `DecompressionParameter`, but it could have been something else.

In any case, it is in a private `_zstd` module, so I don't think it matters a lot.

Happy to change this with Emma's suggestion.

Rogdham replied:
I think you are right, and typing with `type[CompressionParameter]`/`type[DecompressionParameter]` better shows the intent.

I will go with that.

Rogdham commented on stdlib/_zstd.pyi:
```diff
@@ -0,0 +1,95 @@
+
+@final
+class ZstdCompressor:
+    CONTINUE: Final[_ZstdCompressorContinue] = 0
+    FLUSH_BLOCK: Final[_ZstdCompressorFlushBlock] = 1
+    FLUSH_FRAME: Final[_ZstdCompressorFlushFrame] = 2
```
That was to address a suggestion of @emmatyping of making sure the value is only defined once.

Do you want me to revert all occurrences of `_ZstdCompressorContinue`?

e.g. replace `_ZstdCompressorContinue | _ZstdCompressorFlushBlock | _ZstdCompressorFlushFrame` down below with `Literal[0, 1, 2]`

srittau replied:
The rest is fine as is. This was just to point out the redundancy here.

Rogdham commented on stdlib/compression/zstd/_zstdfile.pyi:
```diff
@@ -0,0 +1,110 @@
+_WriteBinaryMode: TypeAlias = Literal["w", "wb", "x", "xb", "a", "ab"]
+_ReadTextMode: TypeAlias = Literal["rt"]
+_WriteTextMode: TypeAlias = Literal["wt", "xt", "at"]
+_PathOrFileBinary: TypeAlias = StrOrBytesPath | IO[bytes]
+_PathOrFileText: TypeAlias = StrOrBytesPath | IO[str]
```
I took this from the `lzma.pyi` file so I thought it was ok. Tell me what you think.

srittau replied:
The lzma stubs probably predate the introduction of protocols and need to be updated as well at some point.

Rogdham commented on stdlib/compression/zstd/_zstdfile.pyi:
```diff
@@ -0,0 +1,110 @@
+    encoding: str | None = ...,
+    errors: str | None = ...,
+    newline: str | None = ...,
+) -> TextIO: ...
```
I took this from the `lzma.pyi` file so I thought it was ok.

This one is straightforward to change though. Do you want me to change it in `lzma.pyi` as well?

srittau replied:
Changing it in lzma as well would be appreciated.

Rogdham commented:
> Also I notice that many defaults are ..., although they have basic defaults in the implementation.

I thought that was implementation details with no use for type checking, but I will change.

---

Likewise, is it needed to define the values for the constants?

For example:
```py
# currently
ZSTD_CLEVEL_DEFAULT: Final[int]

# possible change
ZSTD_CLEVEL_DEFAULT: Final = 3
```

srittau commented:
> Likewise, is it needed to define the values for the constants?

Generally we do that nowadays.

Rogdham commented:
I think I have addressed all your points, feel free to review again, especially the protocols.

I am not sure what the issue with `mypy_primer`, feel free to provide insights if you have any!

emmatyping commented:
> I am not sure what the issue with mypy_primer, feel free to provide insights if you have any!

Definitely looks unrelated. I searched through the codebase and there aren't any references to `compression.zstd`

srittau reviewed (approved):
LGTM, let's see what primer says, but I don't expect any relevant output.
